services/processor/processor/lib/update_from_shotgrid.py

```diff
@@ -0,0 +1,277 @@
+                label=sg_entity_dict["label"],
+                entity_id=sg_entity_dict[CUST_FIELD_CODE_ID],
+                parent_id=ay_parent_entity.id
+            )
+
+        try:
+            self._entity_hub.commit_changes()
+
+            self._sg_session.update(
+                sg_entity_dict["type"],
+                sg_entity_dict[SHOTGRID_ID_ATTRIB],
+                {
+                    CUST_FIELD_CODE_ID: ay_entity.id
+                }
+            )
+        except Exception as e:
+            self.log.error(e)
+            pass
+
+        return ay_entity
+
+    def update_entity(self):
+        """Try to update an entity in Ayon.
+        """
+        sg_entity_dict = get_sg_entity_as_ay_dict(
+            self._sg_session,
+            self._sg_event["entity_type"],
+            self._sg_event["entity_id"],
+        )
+
+        ay_entity = None
+
+        if not sg_entity_dict.get(CUST_FIELD_CODE_ID):
+            self.log.warning("Shotgrid Missing Ayon ID")
+
+            # Attempt to find it by name
+            if sg_entity_dict["type"].lower("task"):
+                ay_entity = self._entity_hub.get_entity_by_name(
+                    sg_entity_dict.get("label")
+                )
+            else:
+                ay_entity = self._entity_hub.get_entity_by_name(
+                    sg_entity_dict.get("name")
+                )
+        else:
+            ay_entity = self._entity_hub.get_entity_by_id(
+                sg_entity_dict.get(CUST_FIELD_CODE_ID)
+            )
+
+        if not ay_entity:
+            self.log.error("Unable to update an non existant entity.")
+            raise ValueError("Unable to update an non existant entity.")
+
+        if int(ay_entity.attribs.get_attribute(SHOTGRID_ID_ATTRIB).value) != int(sg_entity_dict.get(SHOTGRID_ID_ATTRIB)):
+            self.log.error("Missmatching Shotgrid IDs, aborting...")
+            raise ValueError("Missmatching Shotgrid IDs, aborting...")
+
+        if self._sg_event["attribute_name"] in ["code", "name"]:
+            if ay_entity.name != sg_entity_dict["name"]:
+                ay_entity.name = sg_entity_dict["name"]
+
+            if ay_entity.label != sg_entity_dict["label"]:
+                ay_entity.label = sg_entity_dict["label"]
+
+        self._entity_hub.commit_changes()
+
+        if sg_entity_dict.get(CUST_FIELD_CODE_ID) != ay_entity.id:
+            self._sg_session.update(
+                sg_entity_dict["type"],
+                sg_entity_dict[SHOTGRID_ID_ATTRIB],
+                {
+                    CUST_FIELD_CODE_ID: ay_entity.id
+                }
+            )
+
+        return ay_entity
+
+    def remove_entity(self):
+        """Try to remove an entity in Ayon.
+        """
+        sg_entity_dict = get_sg_entity_as_ay_dict(
+            self._sg_session,
+            self._sg_event["entity_type"],
+            self._sg_event["entity_id"],
+        )
+
+        ay_entity = None
```

Comment:
You should not allow to remove entity which has published subsets. You should check `ay_entity.immutable_for_hierarchy` before removement and handle that in some way.